plugin.audio.somafm/resources/lib/addon.py

```diff
@@ -0,0 +1,240 @@
+        ]:
+            value = channel.get_simple_element(element)
+            li.setInfo("Music", {info: value})
+
+        xbmcplugin.addDirectoryItem(
+            handle=handle,
+            url=plugin_url + channel.getid(),
+            listitem=li,
+            totalItems=len(stations),
+        )
+    xbmcplugin.addSortMethod(handle, SORT_METHOD_UNSORTED)
+    xbmcplugin.addSortMethod(handle, SORT_METHOD_LISTENERS)
+    xbmcplugin.addSortMethod(handle, SORT_METHOD_GENRE)
+
+
+def format_priority():
+    setting = xbmcplugin.getSetting(handle, "priority_format")
+    result = [
+        ["mp3"],
+        ["mp3", "aac"],
+        ["aac", "mp3"],
+        ["aac"],
+    ][int(setting)]
+    xbmc.log(
+        f"Format setting is {setting}, using priority {result}", level=xbmc.LOGDEBUG
+    )
+    return result
+
+
+def quality_priority():
+    setting = xbmcplugin.getSetting(handle, "priority_quality")
+    result = [
+        [
+            "slowpls",
+            "fastpls",
+            "highestpls",
+        ],
+        [
+            "fastpls",
+            "slowpls",
+            "highestpls",
+        ],
+        [
+            "fastpls",
+            "highestpls",
+            "slowpls",
+        ],
+        [
+            "highestpls",
+            "fastpls",
+            "slowpls",
+        ],
+    ][int(setting)]
+    xbmc.log(
+        f"Quality setting is {setting}, using priority {result}", level=xbmc.LOGDEBUG
+    )
+    return result
+
+
+def cache_ttl_in_ms():
+    setting = xbmcplugin.getSetting(handle, "cache_ttl")
+    result = [0, __ms_per_day__, 7 * __ms_per_day__, 30 * __ms_per_day__][int(setting)]
+    xbmc.log(f"Cache setting is {setting}, using ttl of {result}", level=xbmc.LOGDEBUG)
+    return result
+
+
+def play(item_to_play):
+    channel_data = fetch_channel_data(
+        fetch_local_channel_data, fetch_remote_channel_data
+    )
+    xml_data = ET.fromstring(channel_data)
+    try:
+        channel_data = xml_data.find(".//channel[@id='" + item_to_play + "']")
+        channel = Channel(
+            handle, tempdir, channel_data, quality_priority(), format_priority()
+        )
+    except:
+        for element in xml_data.findall(".//channel"):
+            channel = Channel(
+                handle, tempdir, element, quality_priority(), format_priority()
+            )
+            if channel.getid() == item_to_play:
+                break
+
+    list_item = ListItem(
+        channel.get_simple_element("title"),
+        channel.get_simple_element("description"),
+        channel.get_content_url(),
+    )
+
+    list_item.setArt(
+        {
+            "icon": channel.geticon(),
+            "thumb": channel.getthumbnail(),
+            "fanart": xbmcvfs.translatePath(
+                "special://home/addons/%s/fanart.jpg" % __addonid__
+            ),
+        }
+    )
+
+    xbmcplugin.setResolvedUrl(handle, True, list_item)
+
+
+def clearcache():
+    shutil.rmtree(tempdir, True)
```

Comment:
This will also delete the user's add-on settings. I don't think that this is convenient?